### PR TITLE
UNIX `remove_dir_all()`: Try recursing first on the slow path

### LINKED ISSUE #94335: Fix for CVE-2022-21658 is unsafe on some file systems
The fix for CVE-2022-21658 (#93112, 54e22eb7dbb615bd44355028d3fd867aa93c0972) appears to have introduced a regression that can cause file system corruption on some UNIX systems.

The new code will attempt to use [unlinkat(2)](https://illumos.org/man/2/unlinkat) in any case where it does not know what type of file a directory entry represents.  Some systems do not provide that information along with the entry names returned while reading a directory, and even on systems which provide a type hint, that hint may be `DT_UNKNOWN`.  In cases where we do not know the type of the entry, the code tries first to unlink it as if it were something other than a directory:

https://github.com/rust-lang/rust/blob/4b043faba34ccc053a4d0110634c323f6c03765e/library/std/src/sys/unix/fs.rs#L1700-L1712

The code assumes that `unlinkat()` on a directory will fail with `EPERM`, or on Linux with the non-standard `EISDIR`, but this is not always the case.  POSIX does suggests `EPERM` as a failure in some, but not all cases:

> [EPERM]
>    The file named by path is a directory, and either the calling process does not have appropriate privileges, or the implementation prohibits using unlink() on directories.

(See [unlink](https://pubs.opengroup.org/onlinepubs/9699919799/functions/unlink.html) in _The Open Group Base Specifications Issue 7, 2018 edition_)

Implementations are not _required_ to prohibit `unlink()` on directories, and indeed UFS file systems on present day illumos and presumably at least some Solaris systems allow `unlink()` of a directory if the user has sufficient privileges.  Other platforms may as well.

Unfortunately unlinking a directory on UFS causes a sort of file system corruption that requires an unmount and a trip through `fsck` to correct.  Any code that uses `std::fs::remove_dir_all()` on UFS has since started causing that kind of corruption, which took a bit of work to track down.

I think the fix is probably relatively simple, and has the benefit of not touching the code path where we believe we know what sort of entry we are removing:

```rust
None => match cvt(unsafe { unlinkat(fd, child.name_cstr().as_ptr(), libc::AT_REMOVEDIR) }) {
    // type unknown - try to rmdir
    Err(err)
        if err.raw_os_error() == Some(libc::EEXIST)
            || err.raw_os_error() == Some(libc::ENOTEMPTY) =>
    {
        // this is a directory that contains files to remove
        remove_dir_all_recursive(Some(fd), Path::new(&child.file_name()))?;
    }
    Err(err) if err.raw_os_error() == Some(libc::ENOTDIR) {
        // this is some other type of entry that we can just unlink
        cvt(unsafe { unlinkat(fd, child.name_cstr().as_ptr(), 0) })?;
    }
    result => {
        result?;
    }
}
```

In short, we should try to `rmdir()` first, rather than `unlink()` first, as that always fails in a safe way that we can detect.

### PR DESCRIPTION
**Pull request by hkratz**
This only affects the _slow_ code path - if there is no `dirent.d_type` or if it is `DT_UNKNOWN`.

POSIX specifies that calling `unlink()` or `unlinkat(..., 0)` on a directory is allowed to succeed:
> The _path_ argument shall not name a directory unless the process has appropriate privileges and the implementation supports using _unlink()_ on directories.

This however can cause dangling inodes requiring an fsck e.g. on Illumos UFS, so we have to avoid that in the common case. We now just try to recurse into it first and unlink() if we can't open it as a directory.

The other two commits integrate the Macos x86-64 implementation reducing redundancy. Split into two commits for better reviewing.

Fixes #94335.

### CODE REVIEW

**Comment by cuviper** [comment]
> On Linux and FreeBSD trying to `unlink()` a directory always fails with EISDIR and EPERM respectively.

Even with NFS to a server OS that does allow unlinking a directory?

**Comment by hkratz** [comment]
> > On Linux and FreeBSD trying to `unlink()` a directory always fails with EISDIR and EPERM respectively.
> 
> Even with NFS to a server OS that does allow unlinking a directory?

According to the man pages it is guaranteed in deviation from the standard POSIX text. I will look into it.

**Comment by hkratz** [comment]
> > On Linux and FreeBSD trying to `unlink()` a directory always fails with EISDIR and EPERM respectively.
> 
> Even with NFS to a server OS that does allow unlinking a directory?

After glossing over the NFSv3/v4 specs it is clear that the OS cannot guarantee that. On the plus side that means we do not need a separate implementation for Linux/FreeBSD and the fast path is still fast anyway.

File: library/std/src/sys/unix/fs.rs [hunk 1]
```diff
@@ -1683,8 +1683,21 @@ mod remove_dir_impl {
...
 
-        // entry is expected to be a directory, open as such
-        let fd = openat_nofollow_dironly(parent_fd, &pcstr)?;
+        // try opening as directory
+        let fd = match openat_nofollow_dironly(parent_fd, &pcstr) {
+            Err(err) if err.raw_os_error() == Some(libc::ENOTDIR) => {
+                // not a directory
+                return match parent_fd {
+                    // return error if this is the root directory
+                    None => return Err(err),
```
Comment (cuviper) [inline library/std/src/sys/unix/fs.rs:1606]: It confused me that we have `return` both on the `match` itself and in this arm. Personally, I'd take it off the `match` and add it into the other arm as well.

**Review by cuviper — CHANGES_REQUESTED** [review]

File: library/std/src/sys/unix/fs.rs [hunk 2]
```diff
@@ -1497,7 +1493,43 @@ mod remove_dir_impl {
...
+            weak!(fn openat(c_int, *const c_char, c_int) -> c_int);
+            openat.get()
+        }
+
+        pub fn has_openat() -> bool {
+            get_openat_fn().is_some()
+        }
+
+        pub unsafe fn openat(dirfd: c_int, pathname: *const c_char, flags: c_int) -> c_int {
+            get_openat_fn().map(|openat| openat(dirfd, pathname, flags)).unwrap_or(libc::ENOSYS)
```
[the remark this answers is not in the document]
Comment (hkratz) [reply library/std/src/sys/unix/fs.rs]: Oops, of course. Thanks for catching that brain-addled bug introduced in a last minute change.

File: library/std/src/sys/unix/fs.rs [hunk 3]
```diff
@@ -1497,7 +1493,43 @@ mod remove_dir_impl {
...
+            weak!(fn fdopendir(c_int) -> *mut DIR, "fdopendir$INODE64");
+            fdopendir.get().map(|fdopendir| fdopendir(fd)).unwrap_or_else(|| {
+                crate::sys::unix::os::set_errno(libc::ENOSYS);
+                crate::ptr::null_mut()
+            })
+        }
+
+        pub unsafe fn unlinkat(dirfd: c_int, pathname: *const c_char, flags: c_int) -> c_int {
+            weak!(fn unlinkat(c_int, *const c_char, c_int) -> c_int);
+            unlinkat.get().map(|unlinkat| unlinkat(dirfd, pathname, flags)).unwrap_or(libc::ENOSYS)
```
[the remark this answers is not in the document]
Comment (hkratz) [reply library/std/src/sys/unix/fs.rs]: See above.

File: library/std/src/sys/unix/fs.rs (hunk 1 above)
Comment (hkratz) [reply library/std/src/sys/unix/fs.rs]: I removed the `return` from the match arm, reordered them and adjusted the comments, if that is OK too.

**Review by cuviper — APPROVED** [review]
Looks good, `r=me` after squashing.

**Comment by hkratz** [comment]
@cuviper Squashed and reworded the macos-impl. commit messages.

**Pull request merged**